tunnel/dns/server.go

```diff
@@ -148,20 +221,76 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
 				A:   address,
 			}
 			msg.Answer = append(msg.Answer, answer)
-		} else {
-			msg.Rcode = dns.RcodeRefused // fail fast, and inspire resolver to query next name server in its list.
+			log.Tracef("response:\n%s\n", msg.String())
+			err := w.WriteMsg(&msg)
+			if err != nil {
+				log.Errorf("write failed: %s", err)
+			}
+			return
 		}
+		if r.upstreamServer != "" {
+			if upstreamResp, err := r.queryUpstream(query); err == nil {
+				err := w.WriteMsg(upstreamResp)
+				if err != nil {
+					log.Errorf("write failed: %s", err)
+				}
+				return
+			}
+		}
+		// No local match or upstream failed
+		r.handleUnanswerable(w, query)
+		return
 	case dns.TypeAAAA:
-		// Always respond with NOERROR for AAAA queries (success, but empty answer)
-		msg.Authoritative = true
-		msg.Rcode = dns.RcodeSuccess
-	}
-	log.Tracef("response:\n%s\n", msg.String())
-	err := w.WriteMsg(&msg)
-	if err != nil {
-		log.Errorf("write failed: %s", err)
+		if _, err := r.getAddress(q.Name); err == nil {
+			msg.Authoritative = true
+			msg.Rcode = dns.RcodeSuccess
+			log.Tracef("response:\n%s\n", msg.String())
+			if err := w.WriteMsg(&msg); err != nil {
+				log.Errorf("write failed: %s", err)
+			}
+			return
+		}
+
+		if r.upstreamServer != "" {
+			if upstreamResp, err := r.queryUpstream(query); err == nil {
+				if err := w.WriteMsg(upstreamResp); err != nil {
+					log.Errorf("write failed: %s", err)
+				}
+				return
+			}
+		}
+
+		r.handleUnanswerable(w, query)
+		return
 	}
 
+	r.handleUnanswerable(w, query)
+}
+
+func (r *resolver) handleUnanswerable(w dns.ResponseWriter, query *dns.Msg) {
+	switch r.unanswered {
+	case unansweredTimeout:
+		log.Tracef("unanswerable query for %s: handling with timeout (no response)", query.Question[0].Name)
+		return
+	case unansweredServfail:
```

Comment:
the only difference between the srvfail and refused paths is the response code. imo it would be better to just set an rcode  variable in the case statement and not duplicate the source code that sends the response.